### DISCUSSION TITLE
Automated Prompt Management across Environments

### DISCUSSION THREAD

**Discussion by jonnyforsterMSG** [langfuse/langfuse#3308] [Support] [answered]
Our setup includes an instance of Langfuse (self hosted) for DEV, UAT and PRD.

Currently when we create or alter prompts it’s a manual task to update this prompt in other environments. This is prone to human error and could result in incidents. A better approach would be to automate the process so it’s consistent. 

Thinking something like - 

Prompts are altered manually only in the DEV environment and only promoted to production in Dev when it’s confirmed to be working correctly. Code refers to the new prompt by version until this point

Prompts are never altered manually in UAT or PRD

Promoting a prompt to production in DEV triggers a pipeline (not sure how) which runs a script removing all prompts in UAT and PRD and replicates what is in DEV via the Api

What are peoples thoughts have you managed to fix this problem another way?

**Comment by marcklingen** [comment, marked as answer]
If you setup a langfuse instance for each environment and want to use langfuse prompt management, you'll need to manage the sync between the instances as you outlined. I'd recommend to use one instance as the root instance and replicate prompts from there to other instances/envs in order to easily deploy prompts from the ui to environments (using langfuse prompt labels).

You'll need to use the API to make these changes on a schedule. The question here is whether you will just push all new prompt versions to the other instances including their labels without needing to remove the previous versions. I think this is the easiest way to replicate prompts.

Steps
- Get all prompts from root instance and mirrors
- For each prompt version that is not on the mirrors already
  - Push the prompt version including its labels to the mirrors
